ecsmeta/ecsmeta.go

```diff
@@ -0,0 +1,118 @@
+package ecsmeta
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/kayac/go-config"
+	"github.com/lestrrat-go/backoff"
+	"github.com/pkg/errors"
+)
+
+func New(opts ...Option) config.DataMap {
+	s := newSetting()
+	for _, opt := range opts {
+		opt(s)
+	}
+	return config.DataMap{
+		"ecsTaskMetadata": getMetadata(s),
+	}
+}
+
+func getMetadata(s *setting) interface{} {
+	b, cancel := s.Start(context.Background())
+	defer cancel()
+	for i := 1; backoff.Continue(b); i++ {
+		metadata, err := getMetadataOnce(s)
+		if err == nil {
+			return metadata
+		}
+		s.Logf("[%d]: unable to get ecs metadata response: %v", i, err)
+	}
+	panic("max retries count reached")
+}
+
+func getMetadataOnce(s *setting) (interface{}, error) {
+	resp, err := s.client.Get(s.endpoint)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get response")
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("incorrect status code %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	var metadata interface{}
+	decoder := json.NewDecoder(resp.Body)
+	if err := decoder.Decode(&metadata); err != nil {
+		return nil, errors.Wrap(err, "unable to decode response body")
+	}
+	return metadata, nil
+}
+
+type Option func(*setting)
+
+type setting struct {
+	client   *http.Client
+	endpoint string
+	policy   backoff.Policy
+	logger   *log.Logger
+}
+
+func newSetting() *setting {
+	endpoint := os.Getenv("ECS_CONTAINER_METADATA_URI") + "/task" //for v3
```

Comment:
v4 endpoint is released recently. https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4.html